##10 jobs, 5 machines with random operation time##

In [2]:
import numpy as np
import random

In [3]:
### Defining functions for make time, crossover & mutation

# To calculate the make time
def cal_makeTime(time_seq):
  len1 = len(time_seq)
  makeTime = 0
  max_list = []
  for i in range(len1):
    opt_time = time_seq[i]
    len2 = len(opt_time)
    for j in range(len2):
      if(i == 0 and j == 0):
        makeTime += opt_time[j]
      elif(i == (len1-1) and j != 0):
        max_list.append(opt_time[j])
        if(j == (len2-1)):
          makeTime += max(max_list)
          max_list = []
      else:
        max_list.append(opt_time[j])
        if(i != 0 and j == 0):
          makeTime += max(max_list)
          max_list = []
  return(makeTime)


# Returns the top x fittest job sequences based on the Make Time
def makeTimeForAll_sequences(seq_list):
  x = []
  for seq in seq_list:
    time_seq = []
    key = ''
    for lst in seq:
      key += lst[0]+" "
      time_seq.append(lst[1])
    makeTime = cal_makeTime(time_seq)
    x.append({key: makeTime})

  return(x)      

def top_fittest_sequences(seq_list, x):

  # Sorting the dictionary items by value to get top x fittest sequences
  sorted_list = sorted(seq_list, key=lambda x: list(x.values())[0])
  # we can use the probability Pi mentioned in step 3 of Algorithm 1 (Project 2 Discussion) but
  # Pi is directly proportional to the Make Time so calcualtion of Pi is redundant

  best_sequence = [(list(d.keys())[0], list(d.values())[0]) for d in sorted_list[:x]]

  return(best_sequence)


# Generates parent pairs from the top fittest parents
def generateParentPairs(topPopulation):
  parentPairs = []
  for i in range(len(topPopulation)):
    for j in range(i+1, len(topPopulation)):
      individual1 = list(topPopulation[i])
      individual2 = list(topPopulation[j])
      parentPairs.append([individual1, individual2])
  return(parentPairs)  

def split_jobList(job_list, len1, len2, len3):
  sublist_1 = job_list[:len1]
  sublist_2 = job_list[len1:len1+len2]
  sublist_3 = job_list[len1+len2:]

  return(sublist_1, sublist_2, sublist_3) 

def getLeftandRight_subList(mainList, subList):
  tempList = mainList[:]
  for i in subList:
    tempList.remove(i)

  x = len(tempList) // 2
  temp1 = tempList[:x]
  temp2 = tempList[x:]

  return (temp1, temp2) 


# generates 45 children by crossover and mutation with 3% probability
def generateChildren(parentPairs, sl1, sl2, sl3 ):
  children = []
  for pairs in parentPairs:     # Performing crossover operation
    parent1 = pairs[0]
    parent2 = pairs[1]
    job_list1 = parent1[0].split()
    job_list2 = parent2[0].split()

    sublist_1len = sl1
    sublist_2len = sl2
    sublist_3len = sl3

    s1, s2, s3 = split_jobList(job_list1, sublist_1len, sublist_2len, sublist_3len)
    s4, s5, s6 = split_jobList(job_list2, sublist_1len, sublist_2len, sublist_3len)

    t1, t2 = getLeftandRight_subList(job_list1, s5)
    t3, t4 = getLeftandRight_subList(job_list2, s2)

    child1 = t3 + s2 + t4
    child2 = t1 + s5 + t2

    if random.random() < 0.03:                    # mutation with a probability of 3%
      i, j = random.sample(range(len(child1)), 2)
      child1[i], child1[j] = child1[j], child1[i]

    children.append(child1)
  return(children)                                                              

In [4]:
jobs = [ 'j1', 'j2', 'j3', 'j4', 'j5', 'j6', 'j7', 'j8', 'j9', 'j10']
initial_jobSeq = {}
for i in range(len(jobs)):
  operation1 = random.randint(5,50)     # Generating random operation times ranging between 5 - 50
  operation2 = random.randint(5,50)
  operation3 = random.randint(5,50)
  operation4 = random.randint(5,50)
  operation5 = random.randint(5,50)
  initial_jobSeq.update({jobs[i]: [operation1, operation2, operation3, operation4, operation5]})

In [5]:
print(initial_jobSeq)       # Structuring the input as a dictionary with keys as the job and values as it's operation times

{'j1': [13, 38, 5, 30, 12], 'j2': [28, 5, 29, 44, 30], 'j3': [8, 37, 13, 11, 11], 'j4': [44, 38, 23, 10, 23], 'j5': [6, 13, 11, 26, 31], 'j6': [6, 25, 50, 6, 30], 'j7': [40, 23, 47, 38, 42], 'j8': [29, 8, 23, 21, 38], 'j9': [19, 35, 10, 20, 12], 'j10': [41, 40, 10, 27, 27]}


In [6]:
for job_id, opt_time in initial_jobSeq.items():
  print(f"{job_id} = {opt_time}")

j1 = [13, 38, 5, 30, 12]
j2 = [28, 5, 29, 44, 30]
j3 = [8, 37, 13, 11, 11]
j4 = [44, 38, 23, 10, 23]
j5 = [6, 13, 11, 26, 31]
j6 = [6, 25, 50, 6, 30]
j7 = [40, 23, 47, 38, 42]
j8 = [29, 8, 23, 21, 38]
j9 = [19, 35, 10, 20, 12]
j10 = [41, 40, 10, 27, 27]


In [7]:
# Initialiing operation times to Jobs that we got from the output in the previous cell
j1 = [13, 38, 5, 30, 12]
j2 = [28, 5, 29, 44, 30]
j3 = [8, 37, 13, 11, 11]
j4 = [44, 38, 23, 10, 23]
j5 = [6, 13, 11, 26, 31]
j6 = [6, 25, 50, 6, 30]
j7 = [40, 23, 47, 38, 42]
j8 = [29, 8, 23, 21, 38]
j9 = [19, 35, 10, 20, 12]
j10 = [41, 40, 10, 27, 27]

In [8]:
# generating 300 random permutaions of the job sequences and adding to the population
population = []
for i in range(300):
    shuffled_data = random.sample(list(initial_jobSeq.items()), len(initial_jobSeq))
    population.append(shuffled_data)

In [9]:
print(population)   # format of the sequence -> ('j8', [9, 35, 11]); 'j8 = job8, [9, 35, 11] = operation times of machine 1, 2 ,3 respectively'
print("Length of the population: ", len(population))

[[('j9', [19, 35, 10, 20, 12]), ('j2', [28, 5, 29, 44, 30]), ('j8', [29, 8, 23, 21, 38]), ('j3', [8, 37, 13, 11, 11]), ('j7', [40, 23, 47, 38, 42]), ('j1', [13, 38, 5, 30, 12]), ('j6', [6, 25, 50, 6, 30]), ('j5', [6, 13, 11, 26, 31]), ('j4', [44, 38, 23, 10, 23]), ('j10', [41, 40, 10, 27, 27])], [('j3', [8, 37, 13, 11, 11]), ('j2', [28, 5, 29, 44, 30]), ('j8', [29, 8, 23, 21, 38]), ('j4', [44, 38, 23, 10, 23]), ('j1', [13, 38, 5, 30, 12]), ('j7', [40, 23, 47, 38, 42]), ('j10', [41, 40, 10, 27, 27]), ('j6', [6, 25, 50, 6, 30]), ('j5', [6, 13, 11, 26, 31]), ('j9', [19, 35, 10, 20, 12])], [('j2', [28, 5, 29, 44, 30]), ('j10', [41, 40, 10, 27, 27]), ('j8', [29, 8, 23, 21, 38]), ('j1', [13, 38, 5, 30, 12]), ('j5', [6, 13, 11, 26, 31]), ('j9', [19, 35, 10, 20, 12]), ('j7', [40, 23, 47, 38, 42]), ('j4', [44, 38, 23, 10, 23]), ('j6', [6, 25, 50, 6, 30]), ('j3', [8, 37, 13, 11, 11])], [('j3', [8, 37, 13, 11, 11]), ('j10', [41, 40, 10, 27, 27]), ('j6', [6, 25, 50, 6, 30]), ('j9', [19, 35, 10, 20

In [10]:
# Function to add generated children to population2
def addChildrenToPopulation(children):
  for child in children:
    population.append([(gene, eval(gene)) for gene in child])                                      

In [11]:
for i in range(100):

  allPopulation_makeTime = makeTimeForAll_sequences(population)   #this calculates the Make time for all the population

  # top = top_fittest_sequences(allPopulation_makeTime, 10)
  topPopulation = top_fittest_sequences(allPopulation_makeTime, 10)  # calculating the top x fittest population based on the Make Time 
                                                        # This function requires population & the top x values to return

  parentPairs = generateParentPairs(topPopulation) # generating parent pairs from the top poulation

  children = generateChildren(parentPairs, 3, 4, 3) # generating childrens from parents, the integer values are the lenghts of substrings

  addChildrenToPopulation(children) # adding children to the initial population

In [12]:
print("Total length of population2 after 100 generations: ", len(population))

Total length of population2 after 100 generations:  4800


In [13]:
print("Top 10 Fittest sequences :", topPopulation)

Top 10 Fittest sequences : [('j6 j10 j1 j2 j7 j4 j8 j5 j9 j3 ', 404), ('j6 j10 j1 j2 j7 j4 j8 j3 j5 j9 ', 404), ('j6 j10 j1 j2 j7 j4 j8 j5 j9 j3 ', 404), ('j6 j10 j1 j2 j7 j4 j8 j9 j5 j3 ', 404), ('j5 j9 j1 j2 j7 j4 j8 j3 j6 j10 ', 404), ('j6 j10 j1 j2 j7 j4 j8 j5 j9 j3 ', 404), ('j6 j10 j1 j2 j7 j4 j8 j5 j9 j3 ', 404), ('j6 j10 j1 j2 j7 j4 j8 j3 j5 j9 ', 404), ('j6 j10 j1 j2 j7 j4 j8 j5 j9 j3 ', 404), ('j6 j10 j1 j2 j7 j4 j8 j9 j5 j3 ', 404)]


There is an interesting finding here. From the top 10 fittest sequences we can see that 9 sequences are of the same sequence.

From the Population in the cell below we can see that the population consists of mix of different sequences but there are cases of duplicate sequences.

This is because I am added the child to the population even if the same child sequence already exits in the population.

*This problem is addressed in the "Additional case" below*

In [18]:
allPopulation_makeTime[500:600]  # This is just to display that the population consists of duplicate job sequences.

[{'j6 j10 j1 j2 j7 j4 j8 j5 j9 j3 ': 404},
 {'j6 j10 j1 j2 j7 j4 j8 j3 j5 j9 ': 404},
 {'j6 j10 j1 j2 j7 j4 j8 j5 j9 j3 ': 404},
 {'j6 j10 j1 j2 j7 j4 j8 j9 j5 j3 ': 404},
 {'j5 j9 j1 j2 j7 j4 j8 j3 j6 j10 ': 404},
 {'j6 j10 j1 j2 j7 j4 j8 j5 j9 j3 ': 404},
 {'j6 j10 j1 j2 j7 j4 j8 j5 j9 j3 ': 404},
 {'j6 j10 j1 j2 j7 j4 j8 j3 j5 j9 ': 404},
 {'j6 j10 j1 j2 j7 j4 j8 j5 j9 j3 ': 404},
 {'j6 j10 j1 j2 j7 j4 j8 j9 j5 j3 ': 404},
 {'j6 j10 j1 j2 j7 j4 j8 j5 j9 j3 ': 404},
 {'j7 j10 j1 j2 j6 j4 j8 j5 j9 j3 ': 438},
 {'j6 j10 j1 j2 j7 j4 j8 j3 j5 j9 ': 404},
 {'j6 j10 j1 j2 j7 j4 j8 j5 j9 j3 ': 404},
 {'j6 j10 j1 j2 j7 j4 j8 j9 j5 j3 ': 404},
 {'j6 j10 j1 j2 j7 j4 j8 j5 j9 j3 ': 404},
 {'j6 j10 j1 j4 j7 j2 j8 j3 j5 j9 ': 412},
 {'j6 j10 j1 j2 j7 j4 j8 j5 j9 j3 ': 404},
 {'j6 j10 j1 j2 j7 j4 j8 j9 j5 j3 ': 404},
 {'j6 j10 j1 j2 j7 j4 j8 j3 j5 j9 ': 404},
 {'j6 j10 j1 j2 j7 j4 j8 j5 j9 j3 ': 404},
 {'j6 j10 j1 j2 j7 j4 j8 j9 j5 j3 ': 404},
 {'j6 j10 j1 j2 j7 j4 j8 j5 j9 j3 ': 404},
 {'j6 j10 j

In [20]:
best_JobSequence = min(topPopulation, key=lambda x: x[1])
print("Job sequence with shortest Make Time: ", best_JobSequence)

Job sequence with shortest Make Time:  ('j6 j10 j1 j2 j7 j4 j8 j5 j9 j3 ', 404)


**The job sequence with minimum Make Time according to GA is 404**

#***Additional case:***

*Preventing addition of the child in population if the child sequence alredy exists in population*

In [22]:
# generating 300 random permutaions of the job sequences and adding to the population
population_noDuplicates = []
for i in range(300):
    shuffled_data = random.sample(list(initial_jobSeq.items()), len(initial_jobSeq))
    population_noDuplicates.append(shuffled_data)

In [27]:
# this functions add the child to population only if child sequence doesn't exist in the population
def addChildrenToPopulation_noDuplicates(children):
  for child in children:
    childSeq = " ".join(child)
    childSeq = childSeq+" "

    for p in allPopulation_makeTime_noDup:
        if childSeq in p:
            break # if child is present in population exit
    else:
        # If child is not present in population add it to the population
        population_noDuplicates.append([(gene, eval(gene)) for gene in child])

In [25]:
print(population_noDuplicates)   # format of the sequence -> ('j8', [9, 35, 11]); 'j8 = job8, [9, 35, 11] = operation times of machine 1, 2 ,3 respectively'
print("Length of the population: ", len(population_noDuplicates))

[[('j7', [40, 23, 47, 38, 42]), ('j4', [44, 38, 23, 10, 23]), ('j6', [6, 25, 50, 6, 30]), ('j8', [29, 8, 23, 21, 38]), ('j5', [6, 13, 11, 26, 31]), ('j2', [28, 5, 29, 44, 30]), ('j10', [41, 40, 10, 27, 27]), ('j3', [8, 37, 13, 11, 11]), ('j9', [19, 35, 10, 20, 12]), ('j1', [13, 38, 5, 30, 12])], [('j2', [28, 5, 29, 44, 30]), ('j10', [41, 40, 10, 27, 27]), ('j7', [40, 23, 47, 38, 42]), ('j4', [44, 38, 23, 10, 23]), ('j1', [13, 38, 5, 30, 12]), ('j8', [29, 8, 23, 21, 38]), ('j5', [6, 13, 11, 26, 31]), ('j6', [6, 25, 50, 6, 30]), ('j3', [8, 37, 13, 11, 11]), ('j9', [19, 35, 10, 20, 12])], [('j4', [44, 38, 23, 10, 23]), ('j7', [40, 23, 47, 38, 42]), ('j1', [13, 38, 5, 30, 12]), ('j9', [19, 35, 10, 20, 12]), ('j8', [29, 8, 23, 21, 38]), ('j2', [28, 5, 29, 44, 30]), ('j3', [8, 37, 13, 11, 11]), ('j5', [6, 13, 11, 26, 31]), ('j10', [41, 40, 10, 27, 27]), ('j6', [6, 25, 50, 6, 30])], [('j7', [40, 23, 47, 38, 42]), ('j1', [13, 38, 5, 30, 12]), ('j8', [29, 8, 23, 21, 38]), ('j10', [41, 40, 10, 2

In [28]:
for i in range(100):

  allPopulation_makeTime_noDup = makeTimeForAll_sequences(population_noDuplicates)   #this calculates the Make time for all the population

  # top = top_fittest_sequences(allPopulation_makeTime, 10)
  topPopulation_noDup = top_fittest_sequences(allPopulation_makeTime_noDup, 10)  # calculating the top x fittest population based on the Make Time 
                                                        # This function requires population & the top x values to return

  parentPairs_noDup = generateParentPairs(topPopulation_noDup) # generating parent pairs from the top poulation

  children_noDup = generateChildren(parentPairs_noDup, 3, 4, 3) # generating childrens from parents, the integer values are the lenghts of substrings

  addChildrenToPopulation_noDuplicates(children_noDup) # adding children to the initial population

In [29]:
print("Total length of population2 after 100 generations: ", len(population_noDuplicates))

Total length of population2 after 100 generations:  469


We can see that if we don't add duplicate child sequences to the population the total population after 100 generations is just 469 (4800 in the previous case)

In [30]:
print("Top 10 Fittest sequences :", topPopulation_noDup)

Top 10 Fittest sequences : [('j6 j9 j5 j1 j2 j10 j7 j4 j3 j8 ', 404), ('j6 j10 j7 j4 j3 j2 j5 j1 j8 j9 ', 404), ('j6 j4 j8 j1 j2 j10 j7 j5 j3 j9 ', 404), ('j5 j8 j3 j1 j2 j10 j7 j6 j4 j9 ', 404), ('j6 j4 j8 j1 j2 j10 j7 j3 j9 j5 ', 404), ('j5 j8 j6 j1 j2 j10 j7 j4 j9 j3 ', 404), ('j6 j5 j8 j1 j2 j10 j7 j4 j9 j3 ', 404), ('j6 j4 j3 j1 j2 j10 j7 j5 j9 j8 ', 404), ('j6 j10 j7 j4 j3 j2 j5 j8 j1 j9 ', 404), ('j6 j7 j4 j2 j10 j1 j3 j5 j9 j8 ', 404)]


In [32]:
best_JobSequence_noDup = min(topPopulation_noDup, key=lambda x: x[1])
print("Job sequence with shortest Make Time: ", best_JobSequence_noDup)

Job sequence with shortest Make Time:  ('j6 j9 j5 j1 j2 j10 j7 j4 j3 j8 ', 404)
